pkg/kypo/auth.go

```diff
@@ -0,0 +1,251 @@
+package kypo
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/http/cookiejar"
+	"net/url"
+	"regexp"
+	"strings"
+	"time"
+)
+
+func (c *Client) signIn() (string, error) {
+	jar, err := cookiejar.New(nil)
+	if err != nil {
+		return "", err
+	}
+
+	httpClient := http.Client{Jar: jar}
+
+	csrf, err := c.authorize(httpClient)
+	if err != nil {
+		return "", err
+	}
+
+	token, csrf, err := c.login(httpClient, csrf)
+	if err != nil {
+		return "", err
+	}
+
+	if token != "" {
+		return token, err
+	}
+
+	return c.authorizeFirstTime(httpClient, csrf)
+}
+
+func (c *Client) authorize(httpClient http.Client) (string, error) {
+	query := url.Values{}
+	query.Add("response_type", "id_token token")
+	query.Add("client_id", c.ClientID)
+	query.Add("scope", "openid email profile")
+	query.Add("redirect_uri", c.Endpoint)
+
+	req, err := http.NewRequest("GET", fmt.Sprintf("%s/csirtmu-dummy-issuer-server/authorize?%s",
+		c.Endpoint, query.Encode()), nil)
+	if err != nil {
+		return "", err
+	}
+
+	res, err := httpClient.Do(req)
+	if err != nil {
+		return "", err
+	}
+	if res.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("authorize failed, got HTTP code: %d", res.StatusCode)
+	}
+
+	body, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return "", err
+	}
+
+	err = res.Body.Close()
+	if err != nil {
+		return "", err
+	}
+
+	csrf, err := extractCsrf(string(body))
+	if err != nil {
+		return "", err
+	}
+
+	return csrf, nil
+}
+
+func extractCsrf(body string) (string, error) {
+	csrfRegex := regexp.MustCompile("<input type=\"hidden\" name=\"_csrf\" value=\"([^\"]+)\" */>")
+	matches := csrfRegex.FindStringSubmatch(body)
+	if len(matches) != 2 {
+		return "", errors.New("failed to match csrf token")
+	}
+	return matches[1], nil
+}
+
+func (c *Client) login(httpClient http.Client, csrf string) (string, string, error) {
+	query := url.Values{}
+	query.Add("username", c.Username)
+	query.Add("password", c.Password)
+	query.Add("_csrf", csrf)
+	query.Add("submit", "Login")
+
+	req, err := http.NewRequest("POST", fmt.Sprintf("%s/csirtmu-dummy-issuer-server/login",
+		c.Endpoint), strings.NewReader(query.Encode()))
+	if err != nil {
+		return "", "", err
+	}
```

Comment:
I think it would be worth naming the return values here. You can save yourself some typing and make the code more clear when returning the values afterwards:
```suggestion
func (c *Client) login(httpClient http.Client, csrf string) (token, csfr string, err error) {
	query := url.Values{}
	query.Add("username", c.Username)
	query.Add("password", c.Password)
	query.Add("_csrf", csrf)
	query.Add("submit", "Login")

	var req *http.Request
	req, err = http.NewRequest("POST", fmt.Sprintf("%s/csirtmu-dummy-issuer-server/login",
		c.Endpoint), strings.NewReader(query.Encode()))
	if err != nil {
		return
	}
```